CORE-8744 Align Cpi related entities and repository with latest JPA guidance

filipesoliveira opened the pull request:
`CpkFileEntity` is now an internal class. Any queries related to this entity should be done using an instance of `CpkFileRepository`.
Added integration tests to cover `CpkFileEntity`.

filipesoliveira committed:
CORE-8744 Align Cpi related entities and repository with latest JPA guidance

CpkFileEntity is now an internal class. Any queries related to this entity should be done using an instance of CpkFileRepository.

vkolomeyko commented on components/virtual-node/cpk-write-service-impl/src/main/kotlin/net/corda/cpk/write/impl/services/db/CpkStorage.kt:
```diff
@@ -1,9 +1,10 @@
 
 interface CpkStorage {
-    fun getCpkIdsNotIn(checksums: List<SecureHash>): List<SecureHash>
+    fun getAllCpkFileIds(fileChecksumsToExclude: List<SecureHash> = emptyList()): List<SecureHash>
```
```suggestion
    fun getAllCpkFileIds(fileChecksumsToExclude: Set<SecureHash> = emptySet()): List<SecureHash>
```

yift-r3 replied:
Why not `Collection<SecureHash>`?

filipesoliveira replied:
I don't mind either suggestion. But out of interest, are we choosing Set over List only for its property of unique elements?

Do we all agree with the generic type, Collection?

vkolomeyko replied:
IMO `Set` better shows the intention of:
- element uniquiness;
- no dependency on the order of the elements being passed in;
- it is best in terms of performance for `contains` like operations.

But `Collection` will also do the trick as a supertype.

filipesoliveira replied:
I see. Thanks Victor.

I have updated the code so that the method uses Collection instead of List to make the code more generic.

vkolomeyko commented on libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/entities/CpkFileEntity.kt:
```diff
@@ -39,7 +38,7 @@ const val QUERY_PARAM_ID = "id"
             " AND f.id = :$QUERY_PARAM_ID"
 )
-class CpkFileEntity(
+internal class CpkFileEntity(
```
Should we also move it to `net.corda.libs.cpi.datamodel.entities.internal` package such that this class not even OSGi exported from this module?

yift-r3 commented on libs/virtual-node/cpi-datamodel/src/integrationTest/kotlin/net/corda/libs/cpi/datamodel/entities/tests/CpkFileRepositoryTest.kt:
```diff
@@ -0,0 +1,49 @@
+class CpkFileRepositoryTest {
+    private val cpkFileRepository = CpkFileRepositoryImpl()
+
+    private lateinit var emFactory: EntityManagerFactory
```
You can simply do `private val emFactory = mock<EntityManagerFactory>()` and remove the setUp...

yift-r3 commented on libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/repository/CpkFileRepositoryImpl.kt:
```diff
@@ -0,0 +1,60 @@
+
+class CpkFileRepositoryImpl: CpkFileRepository {
+    override fun exists(em: EntityManager, cpkChecksum: SecureHash): Boolean {
+        val query = "SELECT count(c) FROM ${CpkFileEntity::class.simpleName} c WHERE c.fileChecksum = :cpkFileChecksum"
```
Why not use the query builder instead of SQL Strings?

filipesoliveira replied:
This code already existed. I did not replaced this SQL string with the query builder to minimised the number of changes and to allow people, and myself, to know from where the logic came from.

filipesoliveira committed:
Moved CpiFileEntity from net.corda.libs.cpi.datamodel.entities to net.corda.libs.cpi.datamodel.entities.internal so the class is not exposed by OSGI.

filipesoliveira committed:
Updated CpkFileRepositoryTest as requested in the code review.

filipesoliveira commented on libs/virtual-node/cpi-datamodel/src/main/kotlin/net/corda/libs/cpi/datamodel/repository/CpkFileRepositoryImpl.kt:
```diff
@@ -0,0 +1,60 @@
+            .singleResult as Long
+
+        // Todos: Should this check be here? It is strange that an exists call throws an exception about too many results found
+        if (entitiesFound > 1) throw NonUniqueResultException("CpkFileEntity with fileChecksum = $cpkChecksum was not unique")
```
Can someone comment on the question in line 17? It is strange to have a method called `exist()` throw an exception.

vkolomeyko replied:
@driessamyn @josephzunigadaly - do you have any recollection why we had this in the first place?

filipesoliveira replied:
The code was copied from "https://github.com/corda/corda-runtime-os/pull/3225/files#diff-1d019fa7f72d0157bfdc313854aacb60f1fe97398a620e5ad10d3f905eb1d20bL53"

filipesoliveira committed:
Replace List for Collection in the interfaces but only for in the arguments of the method.

vkolomeyko commented:
CI reports failure: https://ci02.dev.r3.com/blue/organizations/jenkins/Corda5%2Fcorda-runtime-os/detail/PR-3225/31/tests
for `on putMissingCpk chunks the CPK and publishes it() – net.corda.cpk.write.impl.CpkWriteServiceImplTest` which is likely to be related to the changes made in this PR.

filipesoliveira committed:
Fixed bug in the test "on failing at sub services creation closes the component".

filipesoliveira commented:
> CI reports failure: https://ci02.dev.r3.com/blue/organizations/jenkins/Corda5%2Fcorda-runtime-os/detail/PR-3225/31/tests for `on putMissingCpk chunks the CPK and publishes it() – net.corda.cpk.write.impl.CpkWriteServiceImplTest` which is likely to be related to the changes made in this PR.

Thanks for the heads-up. The test is now fixed.

vkolomeyko reviewed (dismissed):
LGTM

owenstanford commented on libs/virtual-node/cpi-datamodel/src/integrationTest/kotlin/net/corda/libs/cpi/datamodel/entities/tests/CpkFileRepositoryTest.kt:
```diff
@@ -0,0 +1,44 @@
+import net.corda.libs.cpi.datamodel.repository.CpkFileRepositoryImpl
+import net.corda.v5.crypto.SecureHash
+import org.mockito.kotlin.eq
+
```
For the integration tests it would be better to run the repository code against a DB in the same way ```VirtualNodeEntityRepositoryTetst``` does

filipesoliveira replied:
This file is in the wrong place. I will move it to a more appropriate location.

filipesoliveira committed:
Moved the class `CpkFileRepositoryTest` to a more appropriate location.

ronanbrowne reviewed (dismissed):
lgtm (build changes)

vkolomeyko reviewed (dismissed):
LGTM

filipesoliveira committed:
Added integration tasks to cover CpkFileEntity.
Renamed randomChecksumString to genRandomChecksumString.
Renamed randomChecksum to genRandomChecksum.

vkolomeyko commented on libs/virtual-node/cpi-datamodel/src/integrationTest/kotlin/net/corda/libs/cpi/datamodel/entities/tests/CpkFileEntityTest.kt:
```diff
@@ -0,0 +1,242 @@
+    @BeforeEach
+    fun setup() {
+        val inMemoryDbName =  "cpk_file_db_${UUID.randomUUID()}"
+        println("In-memory database name: $inMemoryDbName")
```
If we need to log, can we at least use `logger` rather than outputting it to StdOut without the context information?

filipesoliveira replied:
This is only for testing why the integration test fails when it runs in Jenkins. But it passes in my machine.  The log will be removed afterwards.

owenstanford replied:
DbUtils will return you a Postgres  connection on the CI server as the port env var is set, to recreate the issue locally try running it against a local Postgres

DbUtils code:
```
  val port = System.getProperty(POSTGRES_PORT_PROPERTY)
        return if (!port.isNullOrBlank()) {
            val ds = createPostgresDataSource(dbUser, dbPassword, schemaName, createSchema, rewriteBatchedInserts)
            DbEntityManagerConfiguration(ds, showSql, true, DdlManage.NONE)
        } else {
            logger.info("Using in-memory (HSQL) DB".emphasise())
            TestInMemoryEntityManagerConfiguration(inMemoryDbName, showSql).also {
                if (createSchema) {
                    it.dataSource.connection.createSchema(schemaName)
                }
            }
        }
        ```
```
To use Postgres rather than in-memory (HSQL):

      docker run --rm --name test-instance -e POSTGRES_PASSWORD=password -p 5432:5432 postgres

      gradlew integrationTest -PpostgresPort=5432
```

filipesoliveira replied:
I think I have found my issue. The method `getEntityManagerConfiguration` is a bit misleading. Although the first argument suggests that the database is running in-memory that is not always the case. The behaviour of the method changes based on an environment variable which is a bit odd.

In short, I can have a clean database by changing the name of the database if the database is in-memory. But if a Postgres database is used, I cannot have a fresh database by simply changing the name of the database. Some of my tests rely on the fact the database does not have any data before running the tests.

owenstanford commented on libs/virtual-node/cpi-datamodel/src/integrationTest/kotlin/net/corda/libs/cpi/datamodel/entities/tests/CpkFileEntityTest.kt:
```diff
@@ -0,0 +1,248 @@
+import org.junit.jupiter.api.BeforeEach
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class CpkFileEntityTest {
```
I think this test needs to be renamed? It's testing the CpkFileRepositoryImpl

filipesoliveira replied:
The class has been renamed to CpkFileRepositoryTest.

owenstanford commented on libs/virtual-node/cpi-datamodel/src/test/kotlin/net/corda/libs/cpi/datamodel/entities/tests/CpkFileRepositoryTest.kt:
```diff
@@ -0,0 +1,43 @@
+package net.corda.libs.cpi.datamodel.entities.tests
+
```
This test can be removed now we have an integration test for the CpkFileRepository

filipesoliveira committed:
`CpkFileEntityTest`has been renamed to CpkFileRepositoryTest.

filipesoliveira committed:
Removed the test `CpkFileRepositoryTest`.

filipesoliveira committed:
Removed changes to the build.gradle. No longer required.
